src/docker/image.rs

```diff
@@ -0,0 +1,227 @@
+    type Error = <Self as std::str::FromStr>::Err;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        value.parse()
+    }
+}
+
+impl std::str::FromStr for ImageArchitecture {
+    type Err = eyre::Report;
+    // [os/arch[/variant]=]toolchain
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use serde::de::{
+            value::{Error as SerdeError, StrDeserializer},
+            IntoDeserializer,
+        };
+        if let Some((platform, toolchain)) = s.split_once('=') {
+            let image_toolchain = toolchain.into();
+            let (os, arch, variant) = if let Some((os, rest)) = platform.split_once('/') {
+                let os: StrDeserializer<'_, SerdeError> = os.into_deserializer();
+                let (arch, variant) = if let Some((arch, variant)) = rest.split_once('/') {
+                    let arch: StrDeserializer<'_, SerdeError> = arch.into_deserializer();
+                    (arch, Some(variant))
+                } else {
+                    let arch: StrDeserializer<'_, SerdeError> = rest.into_deserializer();
+                    (arch, None)
+                };
+                (os, arch, variant)
+            } else {
+                eyre::bail!("invalid platform specified")
+            };
+            Ok(ImageArchitecture {
+                architecture: Architecture::deserialize(arch)?,
+                os: Os::deserialize(os)?,
+                variant: variant.map(ToOwned::to_owned),
+                target: image_toolchain,
+            })
+        } else {
+            Ok(ImageArchitecture::from_target(s.into()))
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum Architecture {
+    I386,
+    Amd64,
+    Arm,
+    Arm64,
+    Mips,
+    Mips64,
+    Mips64Le,
+    MipsLe,
+    Ppc64,
+    Ppc64Le,
+    Riscv64,
+    S390x,
+    Wasm,
+}
+
+impl std::fmt::Display for Architecture {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.serialize(f)
+    }
+}
+
+// Supported Oses are on
+// https://rust-lang.github.io/rustup-components-history/aarch64-unknown-linux-gnu.html
+// where rust, rustc and cargo is available (e.g rustup toolchain add works)
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum Os {
+    Darwin,
+    Freebsd,
+    Illumos,
+    Linux,
+    Netbsd,
+    Windows,
+    //Aix,
+    //Android,
+    //Dragonfly,
+    //Ios,
+    //Js,
+    //Openbsd,
+    //Plan9,
+    //Solaris,
+}
+
+impl std::fmt::Display for Os {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.serialize(f)
+    }
+}
+
+impl ImageArchitecture {
+    #[track_caller]
+    pub const fn from_target(target: TargetTriple) -> Self {
+        match target {
+            TargetTriple::Other(_) => todo!(),
```

Comment:
I think we should implement this before we merge, and this should probably return a `Result<Self>`.  If we implement `from_target` for both `Architecture` and `Os`, then this should be pretty much done. Arch is always the first component, and we should be able to implement Os off a match statement on the last 2 values in `triple.split('-')`.